File: scrapegraph-py/scrapegraph_py/local_scraper.py
from pydantic import BaseModel
import requests
from typing import Optional
import json
from .client import ScrapeGraphClient
from .exceptions import raise_for_status_code, APIError

def scrape_text(client: ScrapeGraphClient, website_text: str, prompt: str, schema: Optional[BaseModel] = None) -> str:
    """Scrape and extract structured data from website text using ScrapeGraph AI.

    Args:
        client (ScrapeGraphClient): Initialized ScrapeGraph client
        website_text (str): The text content to analyze
        prompt (str): Natural language prompt describing what data to extract
        schema (Optional[BaseModel]): Pydantic model defining the output structure,
            if provided. The model will be converted to JSON schema before making 
            the request

    Returns:
        str: Extracted data in JSON format matching the provided schema
    """
    endpoint = client.get_endpoint("smartscraper")
    headers = client.get_headers()
    
    payload = {
        "website_text": website_text,
        "user_prompt": prompt
    }
    
    if schema:
        schema_json = schema.model_json_schema()
        payload["output_schema"] = {
            "description": schema_json.get("title", "Schema"),
            "name": schema_json.get("title", "Schema"),
            "properties": schema_json.get("properties", {}),
            "required": schema_json.get("required", [])
        }
    
    try:
        response = requests.post(endpoint, headers=headers, json=payload)
        raise_for_status_code(response.status_code, response)
        return response.text
    except requests.exceptions.RequestException as e:
        raise APIError(f"Request failed: {str(e)}", response=None)

File: scrapegraph-js/src/localScraper.js
import axios from 'axios';
import handleError from './utils/handleError.js';
import { ZodType } from 'zod';
import { zodToJsonSchema } from 'zod-to-json-schema';

/**
 * Extract structured data from local HTML content using ScrapeGraph AI.
 *
 * @param {string} apiKey - The API key for ScrapeGraph AI.
 * @param {string} websiteHtml - HTML content as a string from the local web page to scrape.
 * @param {string} prompt - A natural language description of the data to extract.
 * @param {Object} [schema] - (Optional) Schema object defining the structure of the desired output.
 * @returns {Promise<string>} A JSON string containing the extracted data, formatted to match the schema.
 * @throws {Error} If an HTTP error or validation issue occurs.
 */
export async function localScraper(apiKey, websiteHtml, prompt, schema = null) {
  const endpoint = 'https://api.scrapegraphai.com/v1/localscraper';
  const headers = {
    'accept': 'application/json',
    'SGAI-APIKEY': apiKey,
    'Content-Type': 'application/json',
  };

  const payload = {
    website_html: websiteHtml,
    user_prompt: prompt,
  };

  if (schema) {
    if (schema instanceof ZodType) {
      payload.output_schema = zodToJsonSchema(schema);
    } else {
      throw new Error('The schema must be an instance of a valid Zod schema');
    }
  }

  try {
    const response = await axios.post(endpoint, payload, { headers });
    return response.data;
  } catch (error) {
    handleError(error);
  }
}

/**
 * Retrieve the status or result of a localScraper request, including results of previous requests.
 *
 * @param {string} apiKey - The API key for ScrapeGraph AI.
 * @param {string} requestId - The unique ID associated with the localScraper request.
 * @returns {Promise<string>} A JSON string containing the status or result of the scraping request.
 * @throws {Error} If an error occurs while retrieving the request details.
 */
export async function getLocalScraperRequest(apiKey, requestId) {
  const endpoint = 'https://api.scrapegraphai.com/v1/localscraper/' + requestId;
  const headers = {
    'accept': 'application/json',
    'SGAI-APIKEY': apiKey,
  };

  try {
    const response = await axios.get(endpoint, { headers });
    return response.data;
  } catch (error) {
    handleError(error);
  }
}
